packages/grafana-ui/src/components/Combobox/Combobox.tsx

```diff
@@ -146,11 +148,55 @@ export const Combobox = <T extends string | number>(props: ComboboxProps<T>) =>
         }
       }
 
-      baseSetItems(itemsToSet);
+      const reorderItemsByGroup = (items: Array<ComboboxOption<T>>) => {
+        const reorgItems: Array<ComboboxOption<T> | ComboboxOption<T>['group']> = [];
+        const processedGroups = new Set<ComboboxOption<T>['group']>();
+        const processedItems = new Set<ComboboxOption<T>>();
```

Comment:
I think you could to this in a similar way with Map, and only do a single loop through all the items (no filter or foreach). Then when you've constructed the map, you could do a for...of loop and appends all the items in that manner.